### ISSUE TITLE
DateTime::Duration shouldn't allow non-integer multiplication

### ISSUE THREAD

**Issue by autarch** [houseabsolute/DateTime.pm#25]
Migrated from [rt.cpan.org #117866](https://rt.cpan.org/Ticket/Display.html?id=117866) (status was 'new')

Requestors:
* olaf@wundersolutions.com

From olaf@wundersolutions.com on 2016-09-15 16:06:05:

Multiplying a DateTime::Duration object by a non integer can lead to objects with e.g. 1.5 minutes and 30 seconds, which DateTime::Format::Duration will helpfully render as "1:30"

After discussion with DROLSKY the suggestion was that DateTime::Duration objects probably should die if a non-integer multiplication is attempted. 

See example at http://stackoverflow.com/questions/39499424/adding-mmss-times-dividing-them-and-printing-the-average/39504609#39504609

**Comment by danstewart** [comment]
Is there a preferred way to do this or is something like this alright?
Should I use `Params::Validate` instead?

```perl
unless ($multiplier =~ /^[+-]?\d+$/) {
     die 'Cannot multiply by non integer value';
}

**Comment by autarch** [comment]
You can use `Params::ValidationCompiler`, which is already being used in the code. The type to check would be `t('Int')`.

**Comment by abraxxa** [comment]
The change in version 1.47 from #73 broke one of my apps:

`$avg_duration = $sum_duration->multiply( 1 / $number_of_requests );`

Please advise how to do that now, thanks!

**Comment by autarch** [comment]
I'm not sure what the goal was here. What do you do with `$avg_duration`?

I'm _guessing_ you want to figure out how many seconds something is? In that case you'd get the number of seconds out of the `$sum_duration` and do the division yourself:

```perl
$avg_duration = $sum_duration->in_units('seconds') / $number_of_requests
```

That assumes your duration only contains seconds. If you're generating that duration via DateTime methods, make sure to use the `subtract_datetime_absolute` method, which gives you a duration that only contains seconds and nanoseconds (you can ignore the latter if the source datetime objects are at a 1-second resolution).

**Comment by abraxxa** [comment]
Thanks for the quick response!
I'm passing the $avg_duration object to the template where it's rendered using a custom format_duration method which uses DateTime::Format::Duration and its normalize method.

**Comment by autarch** [comment]
I'm not sure how best to do this now. DT::Duration may not be the right tool here. It's handling of non-integer units was always a bit of a mistake, since DateTime itself can't handle those properly when doing an add or subtract.

Maybe you need a new class that takes a duration and a divisor and does what you need?

**Comment by abraxxa** [comment]
I‘ve worked around it by calculating the average duration in seconds like you suggested and create a new DateTime::Duration object from that to not break the methods API.

**Comment by autarch** [comment]
Fixed in 1.47.
